Mark the volumes as detached when node does not exist

FengyunPan opened the pull request:
If node does not exist, node's volumes will be detached
automatically and become available. So mark them detached and do not return err.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
#50200

**Release note**:
```release-note
NONE
```

FengyunPan commented:
/assign @jingxu97

FengyunPan commented:
/unassign @justinsb 
/unassign @rootfs 
/assign @anguslees 
/assign @dims

dims commented:
/ok-to-test

FengyunPan commented:
/test pull-kubernetes-e2e-kops-aws

dims commented:
/approve

dims commented:
/approve no-issue

jingxu97 commented:
@FengyunPan , could you follow the same pattern as GCE PD or AWS, to have the check in DiskIsAttached function? In that function, if instance is no longer exist, return false without error.

FengyunPan commented:
@jingxu97 Ok, I will check it.

dims commented:
/approve cancel
/lgtm cancel

@jingxu97 can you please remove the do-not-merge?

rootfs commented:
/lgtm cancel

I am not convinced this is an accurate fix. It is safe to detach only if the attachment status is 'attached' (deal with `in-use` and `attaching`)  and the instance id in the attachment cannot be found.

jingxu97 commented:
@rootfs, I think the fix is when node is no longer exist (deleted by cloud provider), we can say the status of the volume is detached. Currently GCE PD and AWS have this logic. Is that the case for cinder?

rootfs commented:
@jingxu97 there is a long list of volume status 
https://developer.openstack.org/api-ref/block-storage/v2/index.html



Status | Description
-- | --
creating | The volume is being created.
available | The volume is ready to attach to an instance.
attaching | The volume is attaching to an instance.
detaching | The volume is detaching from an instance.
in-use | The volume is attached to an instance.
maintenance | The volume is locked and being migrated.
deleting | The volume is being deleted.
awaiting-transfer | The volume is awaiting for transfer.
error | A volume creation error occurred.
error_deleting | A volume deletion error occurred.
backing-up | The volume is being backed up.
restoring-backup | A backup is being restored to the volume.
error_backing-up | A backup error occurred.
error_restoring | A backup restoration error occurred.
error_extending | An error occurred while attempting to extend a volume.
downloading | The volume is downloading an image.
uploading | The volume is being uploaded to an image.
retyping | The volume is changing type to another volume type.
extending | The volume is being extended.

rootfs commented on pkg/volume/cinder/attacher.go:
```diff
@@ -185,6 +186,14 @@ func (attacher *cinderDiskAttacher) VolumesAreAttached(specs []*volume.Spec, nod
+	// So mark them detached.
+	if err == cloudprovider.InstanceNotFound {
+		for _, spec := range specs {
+			volumesAttachedCheck[spec] = false
```
mark it but don't detach it yet.

rootfs replied:
Cannot comment on line 216. 

```go
for volumeID, attached := range attachedResult {
 		if !attached {
 			spec := volumeSpecMap[volumeID]
 			volumesAttachedCheck[spec] = false
 			glog.V(2).Infof("VolumesAreAttached: check volume %q (specName: %q) is no longer attached", volumeID, spec.Name())
 		} else 
               // volume attached, check if the instance is alive here
 	}

FengyunPan replied:
Why we need check if the instance is alive when volume is attached?
Can you tell me more detail message about it? Can you give me a example? Thank you.

rootfs replied:
Checking the instance is alive is not sufficient: the cinder volume could be moved already to other nodes or in a different state. 

We need both attachment info and node info to ensure that the cinder volume is not moved elsewhere and the node the volume is attached is dead, then we can safely detach the volume.

dims replied:
Agree with @rootfs

FengyunPan replied:
Oh, I see. Thank you very much @rootfs.
I update it soon, please take a look.
And I think the change should be inside of openstach_volume.go DiskIsAttached function too. @jingxu97

jingxu97 commented:
Thanks @rootfs 
@FengyunPan since the behavior of Cinder might be different from GCE PD, could you please double check? Also as I mentioned before, the change should be inside of openstach_volume.go DiskIsAttached function.

FengyunPan commented:
@rootfs 
>I am not convinced this is an accurate fix. It is safe to detach only if the attachment status is 'attached' (deal with in-use and attaching) and the instance id in the attachment cannot be found.

Sorry, I can't understand. Can volume be attached to a not exist instance? I think it can't, so it is no need to check it in the case.
Please let me know if my understance is wrong, thank you.

rootfs commented on pkg/volume/cinder/attacher.go:
```diff
@@ -185,7 +186,11 @@ func (attacher *cinderDiskAttacher) VolumesAreAttached(specs []*volume.Spec, nod
 	instanceID, err := attacher.nodeInstanceID(nodeName)
-	if err != nil {
+	if err == cloudprovider.InstanceNotFound {
+		glog.V(2).Infof("VolumesAreAttached: node %q does not exist.", nodeName)
```
print both nodeName and instanceID for better diagnostics.

rootfs replied:
@jingxu97 should raise severity to `glog.Warningf` too?

FengyunPan replied:
@rootfs if instannce is deleted, openstack cloud provider can't get its instanceID, so attacher.nodeInstanceID() return ("", err). Printing nodename is enough.
Yeah, it should use  glog.Warningf, thank you.

dims commented:
@anguslees - any thoughts on this?

jingxu97 commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -412,6 +412,14 @@ func (os *OpenStack) DiskIsAttached(instanceID, volumeID string) (bool, error) {
 
+	if instanceID == "" {
+		if volume.AttachedServerId != "" {
+			return true, nil
```
I don't think the logic here is right. The function is checking whether the volume is attached to the instanced passed in as parameters. Here is the instanceId is empty, it means that the instance is deleted from the cloudprovider. If the volume.AttachedServerId is not empty, that means the volume is a attached to a different instance. But the function should return false because the volume is not attached to the original instance. 

For most of the volume plugins, if the node no longer exist in cloud provider, we can safely assume the volume is detached from the node. But for cinder, could you please double check the behavior if you have the environment? Thanks!

FengyunPan replied:
@jingxu97 IMO, if instanceID is empty, the DiskIsAttached() is checking whether the volume is attached. If the volume is attached, returning true means that the disk is attached, or returning false means that the disk is not attached.
>But the function should return false because the volume is not attached to the original instance.

We can't get the instance id for the original instance which does not exist. If DiskIsAttached() return true when instance does not exist, that means the volume is attached to another instance and we should not mark the volume detached(as rootfs says). If  DiskIsAttached() return false when instance does not exist, that means the volume is available, we can mark the volume detached. The code from line 207 already do it. And that also is my double check, I have tested it.
>But for cinder, could you please double check the behavior if you have the environment? 

I am sorry, I can't understand the meaning of your ‘double check’, can you tell more detail messages? thank you again.

jingxu97 replied:
"We can't get the instance id for the original instance which does not exist. If DiskIsAttached() return true when instance does not exist that means the volume is attached to another instance and we should not mark the volume detached"

DiskIsAttached does not just check disk is attached to ANY node, it checks whether it is attached to the Node that you passed in that function. Suppose disk is currently attached to node B, the function check of DiskIsAttached (disk, nodeA) should return false because it is not attached to nodeA, it is attached to node B. Please let me know if this is clear.

Double check means that I want to confirm that node not found means volume is detached automatically. Seems like this is true for most of cloud provider. But I don't have the cinder volume setup to test it.

FengyunPan replied:
If DiskIsAttached() without my codes, DiskIsAttached() just check whether disk is attached to the instanceID that you passed in that function, and the instanceID can't be empty, right?
Now I add the codes into DiskIsAttached(), DiskIsAttached() can check whether disk is attached to the instanceID(not empty) that you passed in that function, also can check whether disk is attached when instanceID is empty.

Yes, volume will be detached when instance is not found in openstack cinder.

FengyunPan replied:
Is that confused? Is it need to add another func to check whether disk is attached?

jingxu97 replied:
If we are sure that if node not found in openstack cinder indicates that volume is already detached. Then I suggest we can the following changes
 In cinder/attacher.go 
```
if err == cloudprovider.InstanceNotFound {
    //return false with nil error
   ...
}
```
You might have this change in the very first version. Sorry for the confusion. 
@rootfs mentioned that cinder might have a more complicated status, and node not exist might not represent volume is already detached successfully from the node. If this is true, we might not return false directly. But I could not think a good way to check it. Passing an empty string for node instance seems not right because our goal is to check whether the volume is still attached to that specified node (not any other node). What does it mean if node id is empty for this function?

FengyunPan replied:
@jingxu97 @rootfs Yeah, volume will be detached when instance is not found in openstack cinder. I am sorry for that I can't find the doc from openstack, but find the codes:
https://github.com/openstack/nova/blob/master/nova/virt/libvirt/driver.py#L899
nova will detach all the device from instance when user delete instance.
https://github.com/openstack/nova/blob/master/nova/virt/libvirt/driver.py#L983
If instance does not exist, nova will detach the volume too:
https://github.com/openstack/nova/blob/master/nova/virt/libvirt/driver.py#L1389

So we can do: If the node instance is no longer exist (return cloudprovider.InstanceNotFound error), we can safely assume that volume is no longer attached to the node. So DiskIsAttached() return disk is not attached without error.

Commit:
Mark the volumes as detached when node does not exist

If node doesn't exist, OpenStack Nova will assume the volumes
are not attached to it. So mark the volumes as detached and
return false without error.
Fix: #50200

rootfs commented:
Since we are using V2 block storage api, it is safe to say the attachment should be updated with a Nova instance id. This could change in v3 per [info here](https://docs.openstack.org/cinder/pike/contributor/attach_detach_conventions_v2.html). We'll revisit this issue when we pick up v3 block storage sdk 

@kubernetes/sig-openstack-pr-reviews

FengyunPan commented:
@rootfs Yes, I agree. Cinder V3 is different.
But currently we just support cinder v1 and v2. **Can we merge this pr to fix #50200?**
After Merging PR (https://github.com/gophercloud/gophercloud/pull/415), I will add codes to support cinder v3 for openstack cloud provider.  **Then revisit this issue.**

rootfs commented:
/approve

@FengyunPan yes that's my idea too :)

pass to @jingxu97 for lgtm

FengyunPan commented:
@rootfs Yeah, thank you, I will focus on openstack cinder v3 later.

dims commented:
/approve
/lgtm
